lib/internals.js

```diff
@@ -0,0 +1,930 @@
+            );
+          }
+        };
+      }
+      // Do not bother with modules that are only "virtual" and do not have
+      // any files to process.
+      return (await self.apos.asset.computeSourceMeta(options))
+        .filter((entry) => entry.exists);
+    },
+
+    // Generate the import files for all entrypoints and the pre-build manifest.
+    // `suppressErrors` is used to skip errors in the build process.
+    async createImports(entrypoints, suppressErrors = false) {
+      for (const entrypoint of entrypoints) {
+        if (entrypoint.condition === 'nomodule') {
+          self.apos.util.warnDev(
+              `The entrypoint "${entrypoint.name}" is marked as "nomodule". ` +
+              'This is not supported by Vite and will be skipped.'
+          );
+          continue;
+        }
+        if (entrypoint.type === 'bundled') {
+          await self.copyExternalBundledAsset(entrypoint);
+          continue;
+        }
+        const output = await self.getEntrypointOutput(entrypoint, suppressErrors);
+        await self.apos.asset.writeEntrypointFile(output);
+
+        if (!self.entrypointsManifest.some((e) => e.name === entrypoint.name)) {
+          self.entrypointsManifest.push({
+            ...entrypoint,
+            manifest: self.toManifest(entrypoint)
+          });
+        }
+      }
+    },
+
+    // Copy and concatenate the externally bundled assets.
+    async copyExternalBundledAsset(entrypoint) {
+      if (entrypoint.type !== 'bundled') {
+        return;
+      }
+      const filesByOutput = self.apos.asset.getEntrypointManger(entrypoint)
+        .getSourceFiles(self.currentSourceMeta);
+      const manifestFiles = {};
+      for (const [ output, files ] of Object.entries(filesByOutput)) {
+        if (!files.length) {
+          continue;
+        }
+        const raw = files
+          .map(({ path: filePath }) => fs.readFileSync(filePath, 'utf8'))
+          .join('\n');
+
+        await self.apos.asset.writeEntrypointFile({
+          importFile: path.join(self.buildRoot, `${entrypoint.name}.${output}`),
+          prologue: entrypoint.prologue,
+          raw
+        });
+        manifestFiles[output] = manifestFiles[output] || [];
+        manifestFiles[output].push(`${entrypoint.name}.${output}`);
+      }
+      self.entrypointsManifest.push({
+        ...entrypoint,
+        manifest: self.toManifest(entrypoint, manifestFiles)
+      });
+    },
+
+    async getEntrypointOutput(entrypoint, suppressErrors = false) {
+      const manager = self.apos.asset.getEntrypointManger(entrypoint);
+      const files = manager.getSourceFiles(
+        self.currentSourceMeta,
+        { composePath: self.composeSourceImportPath }
+      );
+      const output = await manager.getOutput(files, {
+        modules: self.apos.asset.getRegisteredModules(),
+        suppressErrors
+      });
+      output.importFile = path.join(self.buildRootSource, `${entrypoint.name}.js`);
```

Comment:
can we guarantee that the entrypoint name will not already contain .js at the end, or be something else?